rfcs/20240919-dnncompat/README.md

```diff
@@ -0,0 +1,936 @@
+  integrated into user projects and utilised by auto-translation tools.
+- To deliver an API that closely resembles vendor-specific libraries, enabling
+API calls to be converted primarily through find-and-replace.
+- To handle differences in behaviour between oneDNN and vendor-specific
+libraries internally, thus allowing users to easily convert applications without
+having to debug differences between libraries
+
+In addition, `dnncompat` leverages SYCL Unified Shared Memory (USM) support
+within oneDNN to provide an API that can accept USM pointers as inputs similar
+to vendor-specific library APIs which take pointers to memory as inputs. This
+also allows `dnncompat` to work with memory that is natively allocated, such as
+memory allocated via `cudaMalloc` making integration into native applications
+possible too.
+
+`dnncompat` will be shipped as a set of header files. When oneDNN is built from
+source, these headers will be installed in the directory
+`oneDNN-install-directory/include/dnncompat/`. Users will be able to integrate
+`dnncompat` into their projects by including the oneDNN include directory in
+their include path and adding `#include <dnncompat/dnncompat.hpp>` to their
+project and linking with the oneDNN library.
+
+## SYCLomatic Usage
+
+At present, SYCLomatic provides its own header-only wrapper for oneDNN and
+oneMKL which is used when converting applications. While this approach is
+sufficient, the code produced by SYCLomatic can sometimes be challenging to
+debug and understand for new oneDNN users as it relies on a mixture of custom
+wrappers around oneDNN types along with direct oneDNN API calls. Additionally,
+there are several differences between the supported primitives & their supported
+features in oneDNN and vendor-specific libraries, making debugging incorrect
+output more difficult.
+
+To address these issues, the SYCLomatic headers can be substituted with the
+`dnncompat` wrapper. By using `dnncompat`, the conversion process will be
+simplified, reducing the complexity of the generated code and making it more
+straightforward to understand and debug. This will ensure that developers can
+convert their applications to `dnncompat` & oneDNN more easily, while also
+reducing potential integration issues.
+
+## Scope
+
+Initially the focus will be on a specific set of operators commonly used in the
+VGG-16 model. The following operators will be prioritised: Convolution, BiasAdd,
+Pooling, Activation, Softmax, Local Response Normalization (LRN), and
+TransformTensor (transpose). This will allow for creating a relatively small
+prototype, while showing that `dnncompat` will be useful.
+
+In this initial implementation, these operators will be supported in the forward
+direction. The supported data types initially will be `f32` and `f16`. In the
+future, the support for operations, data types & directions can be expanded
+based on the need of SYCLomatic and community.
+
```

Comment:
I guess a fundamental question we should deal with is: how many version of legacy APIs should we consider? And for how many years should we maintain compatiblity?
In particular, 
- most of the abovementioned functionalities are actually deprecated in cuDNN (see [legacy API ](https://docs.nvidia.com/deeplearning/cudnn/latest/developer/legacy-api.html) documentation and [reference](https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-cnn-library.html)).
- users might be using functionalities that are already no more supported (e.g. see the list for [cuDNN v9](https://docs.nvidia.com/deeplearning/cudnn/latest/api/overview.html#api-changes-900))